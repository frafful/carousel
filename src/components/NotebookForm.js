import React from 'react';
import moment from 'moment';

export default class NotebookForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.notebook ? props.notebook.name : '',
      createdAt: props.notebook ? moment(props.notebook.createdAt) : moment(),
      error: ''
    };
  }

  onSubmit = (e) => {
    e.preventDefault();

    if(!this.state.name) {
      this.setState(() => ({
        error: 'Favor informar o nome do caderno'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        name: this.state.name,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  };

  onNameChange = (e) => {
    const name = e.target.value;
    this.setState(() => ({name}));
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input 
            type="text"
            placeholder="TÃ­tulo"
            value={this.state.name}
            onSubmit={this.onSubmit}
            onChange={this.onNameChange}
          />
          <button>Criar Caderno</button>
        </form>
      </div>
    );
  }
};