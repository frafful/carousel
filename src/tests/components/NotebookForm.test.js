import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import NotebookForm from '../../components/NotebookForm';
import notebooks from '../fixtures/notebooks';

let wrapper;

beforeEach(() => {
  wrapper = shallow(<NotebookForm />);
});

test('should render notebook form correctly', () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render notebook form with notebook data', () => {
  wrapper = shallow(<NotebookForm notebook={notebooks[1]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render error for invalid form submition', () => {
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });

  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should set name on name change', () => {
  const value = 'abc';
  wrapper.find('input').simulate('change', { 
    target: { value }
  });

  expect(wrapper.state('name')).toEqual(value);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should call onSubmit prop for valid form sumission', () => {
  const onSubmit = jest.fn();
  wrapper = shallow(<NotebookForm notebook={notebooks[2]} onSubmit={onSubmit}/>);

  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });

  expect(wrapper.state('error')).toBe('');
  expect(onSubmit).toHaveBeenLastCalledWith({
    name: notebooks[2].name,
    createdAt: notebooks[2].createdAt
  });
});