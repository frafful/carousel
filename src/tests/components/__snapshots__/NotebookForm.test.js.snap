// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render error for invalid form submition 1`] = `
<div>
  <p>
    Favor informar o nome do caderno
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="Título"
      type="text"
      value=""
    />
    <button>
      Criar Caderno
    </button>
  </form>
</div>
`;

exports[`should render notebook form correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="Título"
      type="text"
      value=""
    />
    <button>
      Criar Caderno
    </button>
  </form>
</div>
`;

exports[`should render notebook form with notebook data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="Título"
      type="text"
      value="Festas"
    />
    <button>
      Criar Caderno
    </button>
  </form>
</div>
`;

exports[`should set name on name change 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="Título"
      type="text"
      value="abc"
    />
    <button>
      Criar Caderno
    </button>
  </form>
</div>
`;
